src/ProxyClient/Cloudflare.php

```diff
@@ -0,0 +1,180 @@
+
+    /**
+     * Array of data to send to Cloudflare for purge by URLs request.
+     *
+     * To reduce the number of requests to cloudflare, we buffer the URLs.
+     * During flush, we build requests with batches of URL_BATCH_PURGE_LIMIT.
+     *
+     * @var array<string|array{url: string, headers: string[]}>
+     */
+    private $purgeByUrlsData = [];
+
+    public function __construct(
+        Dispatcher $httpDispatcher,
+        array $options = [],
+        RequestFactory $messageFactory = null
+    ) {
+        if (!function_exists('json_encode')) {
+            throw new \Exception('ext-json is required for cloudflare invalidation');
+        }
+
+        parent::__construct($httpDispatcher, $options, $messageFactory);
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * Tag invalidation only available with Cloudflare enterprise account
+     *
+     * @see https://api.cloudflare.com/#zone-purge-files-by-cache-tags,-host-or-prefix
+     */
+    public function invalidateTags(array $tags)
+    {
+        $this->queueRequest(
+            'POST',
+            sprintf(self::API_ENDPOINT.'/zones/%s/purge_cache', $this->options['zone_identifier']),
+            [],
+            false,
+            $this->json_encode(['tags' => $tags])
+        );
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @see https://api.cloudflare.com/#zone-purge-files-by-url
+     * @see https://developers.cloudflare.com/cache/how-to/purge-cache#purge-by-single-file-by-url For details on headers you can pass to clear the cache correctly
+     */
+    public function purge($url, array $headers = [])
+    {
+        if (!empty($headers)) {
+            $this->purgeByUrlsData[] = [
+                'url' => $url,
+                'headers' => $headers,
+            ];
+        } else {
+            $this->purgeByUrlsData[] = $url;
+        }
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @see https://api.cloudflare.com/#zone-purge-all-files
+     */
+    public function clear()
+    {
+        $this->queueRequest(
+            'POST',
+            sprintf(self::API_ENDPOINT.'/zones/%s/purge_cache', $this->options['zone_identifier']),
+            ['Accept' => 'application/json'],
+            false,
+            $this->json_encode(['purge_everything' => true])
+        );
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc} Queue requests for purge by URLs
+     */
+    public function flush()
+    {
+        // Queue requests for purge by URL
+        foreach (\array_chunk($this->purgeByUrlsData, self::URL_BATCH_PURGE_LIMIT) as $urlChunk) {
+            $this->queueRequest(
+                'POST',
+                sprintf(self::API_ENDPOINT.'/zones/%s/purge_cache', $this->options['zone_identifier']),
+                [],
+                false,
+                $this->json_encode(['files' => $urlChunk])
+            );
+        }
+        $this->purgeByUrlsData = [];
```

Comment:
i added this - we need to clean up this array, in case the invalidation is done from some long running job that periodically flushes.